eventmesh/LockManager.js

```diff
@@ -1,105 +1,184 @@
+                    resourceGroup: <type of resource who is trying to acquire lock ex. backup>
+                    resourceType: <name of resource who is trying to acquire lock ex. defaultbackup>
+                    resourceId: <id of resource who is trying to acquire lock ex.  backup_guid>
+                    operation: <operation of locker ex. update/backup>
+              }
+          })
+      }
+  }
+  Lock is tracked via resource of resource group lock and resource type deploymentlock.
+  id of the resource is instance_id.
+  To lock the deployment we create a new resource of resourceType deploymentlock.
+  If the resource is already present then deployment can't be locked.
   Lock Algorithm looks like the following
-      1. Lock "resource/lock"
-      2. Check "resource/lock/details" value
-      3. if lock value > 1 then lock is already acquired by someone, 
-         unlock "resource/lock" and throw exception.
-      4. else, set the lock details as 
-         {count: 1, operationType: "READ"} 
-      5. release the lock on "resource/lock"
+      1. Create resource lock/deploymentlock/instance_id
+      2. If resource is already present than can't acquire lock
+      3. Check lockdetails from spec.options, if TTL is expired than update the lockdetails
+      4. else can't acquire lock
   */
-  lock(resource, operationType) {
-    const lock = client.lock(resource + CONST.LOCK_KEY_SUFFIX);
-    let lockDetailsChanged = false;
-    return lock.ttl(CONST.LOCK_TTL).acquire()
-      .then(() => {
-        return client.get(resource + CONST.LOCK_DETAILS_SUFFIX).json();
-      })
-      .then(lockDetails => {
-        if (_.get(lockDetails, 'count') > 0) {
-          return lock.release().then(() => {
-            throw new ETCDLockError(`Could not acquire lock for ${resource} as it is already locked.`);
+
+  /**
+   * @param {string} resourceId - Id (name) of the resource that is being locked. In case of deployment lock, it is instance_id
+   * @param {object} lockDetails - Details of the lock that is to be acquired
+   * @param {number} [lockDetails.lockTTL=Infinity] - TTL in miliseconds for the lock that is to be acquired
+   * @param {string} lockDetails.lockType - Type of lock ('READ'/'WRITE')
+   * @param {object} [lockDetails.lockedResourceDetails] - Details of the operation who is trying to acquire the lock
+   * @param {string} [lockDetails.lockedResourceDetails.resourceGroup] - Type of resource for which lock is being acquired. ex: backup
+   * @param {string} [lockDetails.lockedResourceDetails.resourceType] - Name of resource for which lock is being acquired. ex: defaultbackup
+   * @param {string} [lockDetails.lockedResourceDetails.resourceId] - Id of resource for which lock is being acquired. ex: <backup_guid>
+   * @param {string} [lockDetails.lockedResourceDetails.operation=unknown] - Operation type who is acquiring the lock. ex: backup
+   */
+
+  lock(resourceId, lockDetails) {
+    if (!lockDetails) {
```

Comment:
lockDetails should never be empty. Throw error if its sent as such because it would fail in LN90